src/common/configurator.h

```diff
@@ -0,0 +1,158 @@
+// Copyright 2009-2018 Toby Schneider (http://gobysoft.org/index.wt/people/toby)
+//                     GobySoft, LLC (2013-)
+//                     Massachusetts Institute of Technology (2007-2014)
+//                     Community contributors (see AUTHORS file)
+//
+//
+// This file is part of the Goby Underwater Autonomy Project Libraries
+// ("The Goby Libraries").
+//
+// The Goby Libraries are free software: you can redistribute them and/or modify
+// them under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 2.1 of the License, or
+// (at your option) any later version.
+//
+// The Goby Libraries are distributed in the hope that they will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with Goby.  If not, see <http://www.gnu.org/licenses/>.
+
+#ifndef CONFIGURATOR_20181205H
+#define CONFIGURATOR_20181205H
+
+#include "goby/common/configuration_reader.h"
+#include "goby/common/core_helpers.h"
+#include "goby/common/protobuf/app3.pb.h"
+
+namespace goby
+{
+namespace common
+{
+template <typename Config> class ConfiguratorInterface
+{
+  public:
+    void finalize()
+    {
+        if (!config_finalized_)
+        {
+            try
+            {
+                finalize_cfg();
+            }
+            catch (common::ConfigException& e)
+            {
+                handle_config_error(e);
+                throw;
+            }
+            config_finalized_ = true;
+        }
+
+        if (app3_configuration().debug_cfg())
+        {
+            std::cout << cfg().DebugString() << std::endl;
+            exit(EXIT_SUCCESS);
+        }
+    }
+
+    const Config& cfg() const
+    {
+        check_finalized();
+        return const_cfg();
+    }
+
+    const goby::protobuf::App3Config& app3_configuration() const
```

Comment:
It feels a bit confusing having both this and `const_app3_configuration`, especially when they both return a `const goby::protobuf::App3Config&`. Is there a better way to do this, or a better naming convention?

In my opinion the default should be immutable (which it is!) and there should be a special call to access to mutably. This makes it very easy to see which parts of the code are intentionally mutating the configuration. It prevents accidental mutation (which would be a compile-time error) and avoids having code request mutable configuration instances when they aren't needed, which improves optimization opportunities.

Reply:
I want to get away from goby::protobuf::App3Config being a protobuf at all, but rather a C++ struct. When I do this it will be more clear, I hope. Added some TODOs to this effect.